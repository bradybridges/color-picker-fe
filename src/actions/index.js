export const addProjects = (projects) => ({
  type: 'ADD_PROJECTS',
  projects,
});

export const addPalettes = (palettes) => ({
  type: 'ADD_PALETTES',
  palettes,
});
