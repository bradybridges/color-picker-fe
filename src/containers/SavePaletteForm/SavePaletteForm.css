.palette-form {
  text-align: center;
}

.palette-title-input {
  width: 20%;
  background: black;
  color: white;
  height: 30px;
  border-radius: 3px 3px 3px 3px;
  border-color: #2d9fcc;
}

.palette-save-btn {
  margin: 0 auto;
  background: black;
  color: #2d9fcc;
  font-size: 1em;
  border-radius: 3px 3px 3px 3px;
  font-family: 'Permanent Marker';
  border: 1px solid grey;
  width: 20%;
  height: 35px;
  border-color: #2d9fcc;
}

select {
  width: 20%;
  background: black;
  color: white;
  height: 37px;
  border-color: #2d9fcc;
}