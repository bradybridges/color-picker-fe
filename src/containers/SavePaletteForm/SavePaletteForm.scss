.palette-form {
  text-align: center;
}

.palette-title-input {
  background: black;
  border-color: #2d9fcc;
  border-radius: 3px 3px 3px 3px;
  color: white;
  height: 30px;
  margin: 5px;
  width: 20%;
}

.palette-save-btn {
  display: block;
  margin: 0 auto;
  background: black;
  color: #2d9fcc;
  font-size: 1em;
  border-radius: 3px 3px 3px 3px;
  font-family: 'Permanent Marker';
  border: 1px solid grey;
  width: 20%;
  height: 35px;
  border-color: #2d9fcc;
  &:hover {
    cursor: pointer;
  }
}

select {
  background: black;
  border-color: #2d9fcc;
  color: white;
  height: 37px;
  margin: 5px;
  width: 20%;
  &:hover {
    cursor: pointer;
  }
}