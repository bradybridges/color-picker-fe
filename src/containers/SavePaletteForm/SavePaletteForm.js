import React, { Component } from 'react';
import { connect } from 'react-redux';
import { addPalette } from '../../actions/index';
import './SavePaletteForm.css'

export class SavePaletteForm extends Component {
  constructor() {
    super();
    this.state = {
      selectedProject: null
    }
  }

  handleChange = e => {
    this.setState({
      selectedProject: [e.target.value]
    })
  }

  renderProjectOptions = () => {
    const options = this.props.projects.map((project) => {
      return <option key={`${project.name}`} value={project.id}>{project.name}</option>
    });
    return (
      <select onChange={this.handleChange}>
        {options}
      </select>
    );
  }

  handleSubmit = async (e) => {
    e.preventDefault();
    const { tempPalette } = this.props;
    // const projectId = Number(e.target.querySelector('option').value);
    const projectId = parseInt(this.state.selectedProject)
    console.log('projId', projectId)
    // console.log('event target', e.target)
    // console.log('projectId in SavePaletteForm', e.target.querySelector('option'))
    const paletteName = e.target.querySelector('input').value;
    const color1 = tempPalette[0].color;
    const color2 = tempPalette[1].color;
    const color3 = tempPalette[2].color;
    const color4 = tempPalette[3].color;
    const color5 = tempPalette[4].color;
    const response = await this.props.savePalette(projectId, paletteName, color1, color2, color3, color4, color5);
    const paletteId = await response.id;
    const newPalette = { 
      id: paletteId,
      project_id: projectId, 
      palette_name: paletteName, 
      color_1: color1,
      color_2: color2,
      color_3: color3,
      color_4: color4,
      color_5: color5,
    }
    console.log('new palette', newPalette)
    this.props.addPalette(newPalette);
  }

  render() {
    return (
      <div className='form-container'>
      <form className='palette-form' onSubmit={this.handleSubmit}>
        {this.renderProjectOptions()}
        <input className='palette-title-input' placeholder="Enter Palette Name"/>
        <button className='palette-save-btn'>Save Palette</button>
      </form>
      </div>
    );
  }
}

export const mapState = (state) => ({
  tempPalette: state.tempPalette,
  projects: state.projects,
});

export const mapDispatch = (dispatch) => ({
  addPalette: (palette) => dispatch(addPalette(palette)),
});

export default connect(mapState, mapDispatch)(SavePaletteForm);