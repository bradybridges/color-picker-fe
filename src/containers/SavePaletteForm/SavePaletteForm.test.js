import React from 'react';
import { shallow } from 'enzyme';
import { SavePaletteForm, mapState, mapDispatch } from './SavePaletteForm';
import { addPalette } from '../../actions';

describe('SavePaletteForm', () => {
  let wrapper;
  const mockProjects = [
    {
      id: 1,
      name: 'Great Project',
    },
    {
      id: 2,
      name: 'Another Great Project',
    },
  ];
  const mockTempPalette = [
    {
      name: 'color_1',
      color: '#BB2855',
      isLocked: true,
    },
    {
      name: 'color_2',
      color: '#AD12FF',
      isLocked: false,
    },
    {
      name: 'color_3',
      color: '#FFFFFF',
      isLocked: false,
    },
    {
      name: 'color_4',
      color: '#ABABAB',
      isLocked: false,
    },
    {
      name: 'color_5',
      color: '#CCCCCC',
      isLocked: true,
    },
  ];
  const mockSavePalette = jest.fn().mockImplementation(() => {
    return Promise.resolve({ id: 1 });
  });
  const mockAddPalette = jest.fn();
  beforeEach(() => {
    wrapper = shallow(<SavePaletteForm
      projects={mockProjects}
      savePalette={mockSavePalette}
      addPalette={mockAddPalette}
      tempPalette={mockTempPalette}
    />);
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('handleOptionChange should update selected project in state', () => {
    const mockEvent = {
      target: { value: 2 },
    };
    wrapper.instance().handleOptionChange(mockEvent);
    expect(wrapper.state('selectedProject')).toEqual(2);
  });
  it('handleInputChange should update paletteName in state', () => {
    const mockEvent = {
      preventDefault: jest.fn(),
      target: { value: 'New Palette' },
    };
    wrapper.instance().handleInputChange(mockEvent);
    expect(wrapper.state('paletteName')).toEqual('New Palette');
  });
  it('resetForm should reset state', () => {
    const alteredState = { selectedProject: 20, paletteName: 'Test' };
    const expectedState = { selectedProject: 20, paletteName: '' };
    wrapper.setState({ selectedProject: 20, paletteName: 'Test' });
    expect(wrapper.state()).toEqual(alteredState);
    wrapper.instance().resetForm();
    expect(wrapper.state()).toEqual(expectedState);
  });
  it('renderProjectOptions should return a select element with project as options', () => {
    const expected = '{"type":"select","key":null,"ref":null,"props":'
      + '{"children":[{"type":"option","key":"Great Project","ref":null,"props"'
      + ':{"value":1,"children":"Great Project"},"_owner":null,"_store":{}},{"type":"option",'
      + '"key":"Another Great Project","ref":null,"props":{"value":2,"children":"Another Great Project"},'
      + '"_owner":null,"_store":{}}]},"_owner":null,"_store":{}}';
    const result = wrapper.instance().renderProjectOptions();
    expect(JSON.stringify(result)).toEqual(expected);
  });
  it('handleSubmit should update the store, add palette to BE, and reset the form', async () => {
    const mockEvent = { preventDefault: jest.fn() };
    const expected = {
      color_1: '#BB2855',
      color_2: '#AD12FF',
      color_3: '#FFFFFF',
      color_4: '#ABABAB',
      color_5: '#CCCCCC',
      id: 1,
      palette_name: 'Testing',
      project_id: 1,
    };
    wrapper.setState({ paletteName: 'Testing' });
    wrapper.instance().resetForm = jest.fn();
    await wrapper.instance().handleSubmit(mockEvent);
    expect(mockSavePalette).toHaveBeenCalledWith(1, 'Testing', '#BB2855', '#AD12FF', '#FFFFFF', '#ABABAB', '#CCCCCC');
    expect(mockAddPalette).toHaveBeenCalledWith(expected);
    expect(wrapper.instance().resetForm).toHaveBeenCalled();
  });
});

describe('mapState', () => {
  it('should map tempPalette and projects to props', () => {
    const mockTemp = [
      {
        name: 'color_1',
        color: '#FFFFFF',
        isLocked: false,
      },
      {
        name: 'color_2',
        color: '#000000',
        isLocked: false,
      },
      {
        name: 'color_3',
        color: '#CCCCCC',
        isLocked: false,
      },
      {
        name: 'color_4',
        color: '#1F1F1F',
        isLocked: true,
      },
      {
        name: 'color_5',
        color: '#1D1D1D',
        isLocked: true,
      },
    ];
    const mockProjects = [
      {
        id: 1,
        name: 'Great Project',
      },
      {
        id: 2,
        name: 'Another Great Project',
      },
    ];
    const mockState = {
      error: 'Something went wrong',
      tempPalette: mockTemp,
      projects: mockProjects,
    };
    const expectedState = { tempPalette: mockTemp, projects: mockProjects };
    const mappedState = mapState(mockState);
    expect(mappedState).toEqual(expectedState);
  });
});

describe('mapDispatch', () => {
  it('should dispatch addPalette correctly', () => {
    const mockDispatch = jest.fn();
    const mockPalette = {
      id: 1,
      project_id: 1,
      palette_name: 'super dope palette',
      color_1: '#000000',
      color_2: '#ffffff',
      color_3: '#1e1e1e',
      color_4: '#1f1f1f',
      color_5: '#1d1d1d',
    };
    const actionToDispatch = addPalette(mockPalette);
    const mappedDispatch = mapDispatch(mockDispatch);
    mappedDispatch.addPalette(mockPalette);
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });
});
