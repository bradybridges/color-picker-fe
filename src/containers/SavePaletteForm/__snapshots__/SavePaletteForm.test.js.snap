// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SavePaletteForm should match snapshot 1`] = `
<div
  className="form-container"
>
  <form
    className="palette-form"
    onSubmit={[Function]}
  >
    <select
      onChange={[Function]}
    >
      <option
        key="Great Project"
        value={1}
      >
        Great Project
      </option>
      <option
        key="Another Great Project"
        value={2}
      >
        Another Great Project
      </option>
    </select>
    <input
      className="palette-title-input"
      onChange={[Function]}
      placeholder="Enter Palette Name"
      value=""
    />
    <button
      className="palette-save-btn"
      type="submit"
    >
      Save Palette
    </button>
  </form>
</div>
`;
